File: bindings/lua/tests/wallet_manager_test.lua
--[[
Test file for Wallet Manager
]]

local luaunit = require('luaunit')
local WalletManager = require('../satox/wallet/wallet_manager')

TestWalletManager = {}

function TestWalletManager:setUp()
    self.manager = WalletManager.new()
    self.manager:initialize()
end

function TestWalletManager:testInitialization()
    local manager = WalletManager.new()
    luaunit.assertFalse(manager.initialized)
    
    manager:initialize()
    luaunit.assertTrue(manager.initialized)
    
    -- Test double initialization
    luaunit.assertError(function()
        manager:initialize()
    end, "WalletManager already initialized")
end

function TestWalletManager:testCreateWallet()
    -- Test successful wallet creation
    local wallet = self.manager:createWallet("test_wallet", "password123")
    luaunit.assertNotNil(wallet)
    luaunit.assertEquals(wallet.name, "test_wallet")
    luaunit.assertNotNil(wallet.address)
    luaunit.assertNotNil(wallet.publicKey)
    luaunit.assertNotNil(wallet.privateKey)
    luaunit.assertNotNil(wallet.created)
    luaunit.assertNotNil(wallet.lastUsed)
    luaunit.assertEquals(wallet.version, "1.0.0")
    
    -- Test duplicate wallet name
    luaunit.assertError(function()
        self.manager:createWallet("test_wallet", "password123")
    end, "Wallet already exists: test_wallet")
    
    -- Test invalid wallet name
    luaunit.assertError(function()
        self.manager:createWallet(nil, "password123")
    end, "Invalid wallet name")
    
    -- Test invalid password
    luaunit.assertError(function()
        self.manager:createWallet("test_wallet2", nil)
    end, "Invalid password")
end

function TestWalletManager:testImportWallet()
    -- Test successful wallet import
    local privateKey = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
    local wallet = self.manager:importWallet("imported_wallet", privateKey, "password123")
    luaunit.assertNotNil(wallet)
    luaunit.assertEquals(wallet.name, "imported_wallet")
    luaunit.assertNotNil(wallet.address)
    luaunit.assertNotNil(wallet.publicKey)
    luaunit.assertNotNil(wallet.privateKey)
    
    -- Test invalid private key
    luaunit.assertError(function()
        self.manager:importWallet("test_wallet", "invalid_key", "password123")
    end, "Invalid private key format")
    
    -- Test duplicate wallet name
    luaunit.assertError(function()
        self.manager:importWallet("imported_wallet", privateKey, "password123")
    end, "Wallet already exists: imported_wallet")
end

function TestWalletManager:testGetWallet()
    -- Create a test wallet
    self.manager:createWallet("test_wallet", "password123")
    
    -- Test successful wallet retrieval
    local wallet = self.manager:getWallet("test_wallet")
    luaunit.assertNotNil(wallet)
    luaunit.assertEquals(wallet.name, "test_wallet")
    
    -- Test non-existent wallet
    luaunit.assertError(function()
        self.manager:getWallet("non_existent")
    end, "Wallet not found: non_existent")
end

function TestWalletManager:testListWallets()
    -- Create test wallets
    self.manager:createWallet("wallet1", "password123")
    self.manager:createWallet("wallet2", "password123")
    
    -- Test wallet listing
    local wallets = self.manager:listWallets()
    luaunit.assertNotNil(wallets)
    luaunit.assertEquals(#wallets, 2)
    luaunit.assertNotNil(wallets["wallet1"])
    luaunit.assertNotNil(wallets["wallet2"])
end

function TestWalletManager:testDeleteWallet()
    -- Create a test wallet
    self.manager:createWallet("test_wallet", "password123")
    
    -- Test successful wallet deletion
    self.manager:deleteWallet("test_wallet", "password123")
    luaunit.assertError(function()
        self.manager:getWallet("test_wallet")
    end, "Wallet not found: test_wallet")
    
    -- Test deletion with wrong password
    self.manager:createWallet("test_wallet2", "password123")
    luaunit.assertError(function()
        self.manager:deleteWallet("test_wallet2", "wrong_password")
    end, "Invalid password")
end

function TestWalletManager:testSignTransaction()
    -- Create a test wallet
    self.manager:createWallet("test_wallet", "password123")
    
    -- Test transaction signing
    local transaction = { to = "0x123", value = 100 }
    local signature = self.manager:signTransaction("test_wallet", transaction, "password123")
    luaunit.assertNotNil(signature)
    
    -- Test signing with wrong password
    luaunit.assertError(function()
        self.manager:signTransaction("test_wallet", transaction, "wrong_password")
    end, "Invalid password")
end

function TestWalletManager:testVerifySignature()
    -- Create a test wallet
    self.manager:createWallet("test_wallet", "password123")
    
    -- Test signature verification
    local transaction = { to = "0x123", value = 100 }
    local signature = self.manager:signTransaction("test_wallet", transaction, "password123")
    local wallet = self.manager:getWallet("test_wallet")
    
    local isValid = self.manager:verifySignature(transaction, signature, wallet.publicKey)
    luaunit.assertTrue(isValid)
end

function TestWalletManager:testChangePassword()
    -- Create a test wallet
    self.manager:createWallet("test_wallet", "old_password")
    
    -- Test password change
    self.manager:changePassword("test_wallet", "old_password", "new_password")
    
    -- Verify old password no longer works
    luaunit.assertError(function()
        self.manager:signTransaction("test_wallet", {}, "old_password")
    end, "Invalid password")
    
    -- Verify new password works
    local signature = self.manager:signTransaction("test_wallet", {}, "new_password")
    luaunit.assertNotNil(signature)
end

function TestWalletManager:testBackupAndRestore()
    -- Create a test wallet
    self.manager:createWallet("test_wallet", "password123")
    
    -- Test wallet backup
    local backup = self.manager:backupWallet("test_wallet", "password123")
    luaunit.assertNotNil(backup)
    luaunit.assertEquals(backup.name, "test_wallet")
    luaunit.assertNotNil(backup.address)
    luaunit.assertNotNil(backup.publicKey)
    luaunit.assertNotNil(backup.privateKey)
    
    -- Delete original wallet
    self.manager:deleteWallet("test_wallet", "password123")
    
    -- Test wallet restore
    local restored = self.manager:restoreWallet(backup, "password123")
    luaunit.assertNotNil(restored)
    luaunit.assertEquals(restored.name, "test_wallet")
    luaunit.assertEquals(restored.address, backup.address)
    luaunit.assertEquals(restored.publicKey, backup.publicKey)
end

function TestWalletManager:testExportWallet()
    -- Create a test wallet
    self.manager:createWallet("test_wallet", "password123")
    
    -- Test wallet export
    local exported = self.manager:exportWallet("test_wallet", "password123")
    luaunit.assertNotNil(exported)
    luaunit.assertEquals(exported.name, "test_wallet")
    luaunit.assertNotNil(exported.address)
    luaunit.assertNotNil(exported.publicKey)
    luaunit.assertNotNil(exported.privateKey)
    
    -- Test export with wrong password
    luaunit.assertError(function()
        self.manager:exportWallet("test_wallet", "wrong_password")
    end, "Invalid password")
end

-- Run tests
os.exit(luaunit.LuaUnit.run()) 

File: bindings/python/tests/wallet_manager_test.py
"""
Satox SDK Wallet Manager Tests
"""

import unittest
from satox.wallet.wallet_manager import WalletManager, WalletConfig, NetworkType, Wallet

class TestWalletManager(unittest.TestCase):
    """Test cases for Wallet Manager"""

    def setUp(self):
        """Set up test environment"""
        self.manager = WalletManager()
        self.manager.initialize()

    def test_initialize(self):
        """Test Wallet Manager initialization"""
        # Test default initialization
        manager = WalletManager()
        self.assertFalse(manager.initialized)
        
        manager.initialize()
        self.assertTrue(manager.initialized)

        # Test custom configuration
        config = WalletConfig(
            version="2.0.0",
            min_password_length=10,
            max_wallet_name_length=40,
            min_wallet_name_length=5
        )
        manager = WalletManager()
        manager.initialize(config)
        self.assertEqual(manager.config.version, "2.0.0")
        self.assertEqual(manager.config.min_password_length, 10)
        self.assertEqual(manager.config.max_wallet_name_length, 40)
        self.assertEqual(manager.config.min_wallet_name_length, 5)

        # Test double initialization
        with self.assertRaises(RuntimeError):
            manager.initialize()

    def test_create_wallet(self):
        """Test wallet creation"""
        # Test valid wallet creation
        wallet = self.manager.create_wallet("test_wallet", "password123")
        self.assertEqual(wallet.name, "test_wallet")
        self.assertEqual(wallet.network, NetworkType.MAINNET)
        self.assertTrue(wallet.private_key.startswith("0x"))
        self.assertTrue(wallet.public_key.startswith("0x"))
        self.assertTrue(wallet.address.startswith("0x"))

        # Test duplicate wallet
        with self.assertRaises(RuntimeError):
            self.manager.create_wallet("test_wallet", "password123")

        # Test invalid wallet name
        with self.assertRaises(ValueError):
            self.manager.create_wallet("test@wallet", "password123")

        # Test invalid password
        with self.assertRaises(ValueError):
            self.manager.create_wallet("test_wallet2", "short")

    def test_import_wallet(self):
        """Test wallet import"""
        # Test valid wallet import
        private_key = "0x" + "1" * 64
        wallet = self.manager.import_wallet("imported_wallet", private_key, "password123")
        self.assertEqual(wallet.name, "imported_wallet")
        self.assertEqual(wallet.private_key, private_key)
        self.assertEqual(wallet.network, NetworkType.MAINNET)

        # Test duplicate wallet
        with self.assertRaises(RuntimeError):
            self.manager.import_wallet("imported_wallet", private_key, "password123")

        # Test invalid private key
        with self.assertRaises(ValueError):
            self.manager.import_wallet("invalid_wallet", "invalid_key", "password123")

    def test_get_wallet(self):
        """Test wallet retrieval"""
        # Create test wallet
        self.manager.create_wallet("test_wallet", "password123")

        # Test valid wallet retrieval
        wallet = self.manager.get_wallet("test_wallet")
        self.assertEqual(wallet.name, "test_wallet")

        # Test non-existent wallet
        with self.assertRaises(RuntimeError):
            self.manager.get_wallet("non_existent")

    def test_list_wallets(self):
        """Test wallet listing"""
        # Create test wallets
        self.manager.create_wallet("wallet1", "password123")
        self.manager.create_wallet("wallet2", "password123")

        # Test wallet listing
        wallets = self.manager.list_wallets()
        self.assertEqual(len(wallets), 2)
        self.assertIn("wallet1", wallets)
        self.assertIn("wallet2", wallets)

    def test_delete_wallet(self):
        """Test wallet deletion"""
        # Create test wallet
        self.manager.create_wallet("test_wallet", "password123")

        # Test valid wallet deletion
        self.manager.delete_wallet("test_wallet", "password123")
        with self.assertRaises(RuntimeError):
            self.manager.get_wallet("test_wallet")

        # Test non-existent wallet deletion
        with self.assertRaises(RuntimeError):
            self.manager.delete_wallet("non_existent", "password123")

    def test_sign_transaction(self):
        """Test transaction signing"""
        # Create test wallet
        wallet = self.manager.create_wallet("test_wallet", "password123")

        # Test transaction signing
        transaction = {"data": "test"}
        signature = self.manager.sign_transaction("test_wallet", transaction, "password123")
        self.assertTrue(signature.startswith("0x"))

        # Test non-existent wallet
        with self.assertRaises(RuntimeError):
            self.manager.sign_transaction("non_existent", transaction, "password123")

    def test_verify_signature(self):
        """Test signature verification"""
        # Create test wallet
        wallet = self.manager.create_wallet("test_wallet", "password123")

        # Test signature verification
        transaction = {"data": "test"}
        signature = self.manager.sign_transaction("test_wallet", transaction, "password123")
        self.assertTrue(self.manager.verify_signature(transaction, signature, wallet.public_key))

        # Test invalid signature
        self.assertFalse(self.manager.verify_signature(transaction, "0xinvalid", wallet.public_key))

    def test_change_password(self):
        """Test password change"""
        # Create test wallet
        self.manager.create_wallet("test_wallet", "password123")

        # Test password change
        self.manager.change_password("test_wallet", "password123", "new_password123")

        # Test non-existent wallet
        with self.assertRaises(RuntimeError):
            self.manager.change_password("non_existent", "password123", "new_password123")

    def test_backup_wallet(self):
        """Test wallet backup"""
        # Create test wallet
        self.manager.create_wallet("test_wallet", "password123")

        # Test wallet backup
        backup = self.manager.backup_wallet("test_wallet", "password123")
        self.assertEqual(backup.name, "test_wallet")
        self.assertTrue(isinstance(backup, Wallet))

        # Test non-existent wallet
        with self.assertRaises(RuntimeError):
            self.manager.backup_wallet("non_existent", "password123")

    def test_restore_wallet(self):
        """Test wallet restoration"""
        # Create and backup test wallet
        self.manager.create_wallet("test_wallet", "password123")
        backup = self.manager.backup_wallet("test_wallet", "password123")

        # Delete original wallet
        self.manager.delete_wallet("test_wallet", "password123")

        # Test wallet restoration
        restored = self.manager.restore_wallet(backup, "password123")
        self.assertEqual(restored.name, "test_wallet")
        self.assertEqual(restored.private_key, backup.private_key)

        # Test duplicate wallet
        with self.assertRaises(RuntimeError):
            self.manager.restore_wallet(backup, "password123")

    def test_export_wallet(self):
        """Test wallet export"""
        # Create test wallet
        self.manager.create_wallet("test_wallet", "password123")

        # Test wallet export
        exported = self.manager.export_wallet("test_wallet", "password123")
        self.assertEqual(exported.name, "test_wallet")
        self.assertTrue(isinstance(exported, Wallet))

        # Test non-existent wallet
        with self.assertRaises(RuntimeError):
            self.manager.export_wallet("non_existent", "password123")

if __name__ == "__main__":
    unittest.main() 